src/server/pps/server/api_server.go

```diff
@@ -458,32 +524,93 @@ func (a *apiServer) Version(version int64) error {
 	return nil
 }
 
-func (a *apiServer) AddShard(shard uint64, version int64) error {
-	pipelineInfos, err := a.ListPipeline(context.Background(), &ppsclient.ListPipelineRequest{})
+func (a *apiServer) AddShard(shard uint64) error {
+	persistClient, err := a.getPersistClient()
 	if err != nil {
 		return err
 	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	a.shardCancelFuncsLock.Lock()
+	defer a.shardCancelFuncsLock.Unlock()
+	if _, ok := a.shardCancelFuncs[shard]; ok {
+		return fmt.Errorf("shard %d is being added twice; this is likely a bug", shard)
+	}
+	a.shardCancelFuncs[shard] = cancel
+
+	client, err := persistClient.SubscribeNewPipeline(ctx, &persist.SubscribeNewPipelineRequest{})
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		pipelineChan := make(chan *ppsclient.PipelineInfo)
+		go func() {
+			for {
+				pipelineInfo, err := client.Recv()
+				if err != nil {
+					protolion.Printf("error receiving new pipeline: %v", err)
+				}
+				pipelineChan <- newPipelineInfo(pipelineInfo)
+			}
+		}()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case pipelineInfo := <-pipelineChan:
+				if a.hasher.HashPipeline(pipelineInfo.Pipeline) == shard {
+					go func() {
+						if err := a.runPipeline(pipelineInfo); err != nil {
+							protolion.Printf("error running pipeline: %v", err)
+						}
+					}()
+				}
+			}
+		}
+	}()
+
+	// If we did the following before we subscribe to changes, a new pipeline
+	// could come in after we did the following but before we subscribe to
+	// changes, so we might miss that pipeline.
```

Comment:
I think you can avoid needing to calls by passing the flag `include_initial` to the rethinkdb query.